qcodes/instrument_drivers/Keysight/keysight_e4980a.py

```diff
@@ -386,6 +385,32 @@ def _set_measurement(self,
         self._measurement_pair = measurement_pair
         self.write(f":FUNCtion:IMPedance {measurement_pair.name}")
 
+    def _get_voltage_level(self) -> float:
+        """
+        Gets voltage level if signal is set with voltage level parameter
+        otherwise raises an error.
+        """
+        try:
```

Comment:
Would it make sense to check the cached value of the `current_level` parameter before trying to set the voltage level parameter?

Reply:
That seems reasonable, a unit test of this would be good too if this is added.

Reply:
I though about checking cached value of the `current_level` param but we cannot raise an error in that case because otherwise if say signal is set using current level param and someone tries to set now voltage level he will never be able to do so with error raise in place. Though we can log a warning in this scenario.

Reply:
I am not sure I understand. 

I expected the work flow to be:

```
set current level none zero
do some measurements
set current level back to zero
set voltage level none zero
do some other measurements
```

And then in case the user forgets to do `set current level back to zero` an error will be raised when the user tries to `set voltage level none zero`

Reply:
We can restrict the instrument to the workflow which you have mentioned with  a check but instrument allows to set the signal using voltage level if set with current level previously.

Reply:
I will check again with instrument and make the changes.

Reply:
Ok could I make another suggestion?

Add a manual parameter that does not communicate with the instrument that keeps track of which mode we are in.
By default this will be unset. 

When setting the current_level to something none zero this will set the mode to current mode and the other way around if setting voltage level.

Now if you try to get the voltage while in current mode we can raise an error and the other way around. 

If we assume that users are not touching the front panel will this not work?

Reply:
Yes, this will work perfectly. Thanks for the suggestion Jens :)